File: Lectures/Lecture 1/linked_list.py

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

    def __str__(self):
        return self.data


class LinkedList:
    def __init__(self):
        self.head = None

    def __str__(self):
        current_node = self.head
        my_str = ""

        while current_node is not None:
            my_str += f"{current_node.data} -> "
            current_node = current_node.next

        my_str += "None"
        return my_str

    # count nodes in the list
    def countNodes(self):
        temp = self.head
        i = 0
        while(temp != None):
            i += 1
            temp = temp.next
        return i

    # Counts the total sum of node
    def sumNodes(self):
        current_node = self.head
        sumCount = 0

        if(self.head is None):
            print("No value given")
            return
        while(current_node is not None):
            sumCount += current_node.data
            current_node = current_node.next
        return sumCount

    # Add new element at the end of the list
    def append(self, newParam):
        newNode = Node(newParam)
        if(self.head is not None):
            temp = self.head
            while(temp.next != None):
                temp = temp.next
            temp.next = newNode
        else:
            if(self.head == None):
                self.head = newNode
                return newNode


if __name__ == '__main__':
    llist = LinkedList()
    # print(llist)

    """box1 = Node("a")
    llist.head = box1"""
    llist.append(1)
    print(llist)
    
    """box2 = Node("b")
    box1.next = box2"""
    llist.append(2)
    print(llist)

    """box3 = Node("c")
    box2.next = box3"""
    llist.append(3)
    print(llist)

    """box4 = Node("d")
    box3.next = box4"""
    llist.append(4)
    print(llist)

    """box5 = Node("e")
    box4.next = box5"""
    llist.append(5)
    print(llist)

    current_node = llist.head
    # Börja på första noden och gå till next, tills vi kommer till slutet
    while current_node is not None:
        # Skriva ut vad noden innehåller
        print(current_node.data)

        # Gå vidare till nästa nod
        current_node = current_node.next

    print("No. of nodes:", llist.countNodes())
    print("Sum amount of nodes:", llist.sumNodes())
    

File: Lectures/Lecture 1/linked_list.cpp
#include <iostream>

using namespace std;

class Node
{
public:
    int data;
    Node *next;

    // Default constructor
    Node()
    {
        data = 0;
        next = NULL;
    }

    // Constructor with data
    Node(int data)
    {
        this->data = data;
        this->next = NULL;
    }

};

class LinkedList
{
public:
    Node *head, *tail;

    LinkedList()
    {
        head = NULL;
        tail = NULL;
    }

    void printList();
    void append(int);
    int count();
    int sum();

};

void LinkedList::append(int value)
{
    Node *nn = new Node(value);

    if (this->head == NULL)
    {
        this->head = nn;
        this->tail = nn;
    }
    else 
    {
        this->tail->next = nn;
        this->tail = nn;
    }
}

int LinkedList::count()
{
    Node *node = head;
    int count = 0;

    while(node!=NULL)
    {
        count++;
        node = node->next;
    }

    return count;
}

int LinkedList::sum()
{
    Node *node = head;
    int sum = 0;

    while (node!=NULL)
    {
        sum+= node->data;
        node = node->next;
    }

    return sum;    
}
void LinkedList::printList()
{
    Node *temp = head;

    // Check for empty list.
    if (head == NULL)
    {
        cout << "List empty" << endl;
        return;
    }
    // Traverse the list.
    while (temp != NULL)
    {
        cout << temp->data << " -> ";
        temp = temp->next;
    }
    cout << "None" << endl;
}

int main()
{
    LinkedList llist;
    llist.printList();

    llist.append(1);
    llist.printList();

    llist.append(6);
    llist.printList();

    llist.append(3);
    llist.printList();

    Node *currentNode = llist.head;
    // Börja på första noden och gå till next, tills vi kommer till slutet
    while (currentNode != NULL) {
        cout << currentNode->data << endl;

        // Gå vidare till nästa nod
        currentNode = currentNode->next;
    }
    
    cout << "Count sum: " << llist.count() << endl;
    cout << "Data sum: " << llist.sum() << endl;
}